webhook/pr_review.go

```diff
@@ -0,0 +1,384 @@
+			time.Sleep(time.Duration(sleepTimeSecond) * time.Second)
+			builder.Reset()
+		}
+	}
+}
+
+func getChatGPTComment(patch string) (string, error) {
+	client, err := openai.New(
+		openai.WithToken(viper.GetString("openai.api_key")),
+		openai.WithModel(viper.GetString("openai.model")),
+		openai.WithOrgID(viper.GetString("openai.org_id")),
+		openai.WithProxyURL(viper.GetString("openai.proxy")),
+		openai.WithSocksURL(viper.GetString("openai.socks")),
+		openai.WithBaseURL(viper.GetString("openai.base_url")),
+		openai.WithTimeout(viper.GetDuration("openai.timeout")),
+		openai.WithMaxTokens(viper.GetInt("openai.max_tokens")),
+		openai.WithTemperature(float32(viper.GetFloat64("openai.temperature"))),
+	)
+	if err != nil {
+		log.Println("Error creating OpenAI client:", err)
+		return "", err
+	}
+
+	out, err := util.GetTemplateByString(
+		prompt.CodeReviewTemplate,
+		util.Data{
+			"file_diffs": patch,
+		},
+	)
+	if err != nil {
+		log.Println("Error creating prompt:", err)
+		return "", err
+	}
+
+	ctx := context.Background()
+	resp, err := client.Completion(ctx, out)
+	if err != nil {
+		log.Println("Error completing prompt:", err)
+		return "", err
+	}
+	summarizeMessage := resp.Content
+
+	log.Println("PromptTokens: " + strconv.Itoa(resp.Usage.PromptTokens) +
+		", CompletionTokens: " + strconv.Itoa(resp.Usage.CompletionTokens) +
+		", TotalTokens: " + strconv.Itoa(resp.Usage.TotalTokens))
+
+	// Output core review summary
+	log.Println("================Review Summary====================")
+	log.Println("\n" + strings.TrimSpace(summarizeMessage) + "\n\n")
+	log.Println("==================================================")
+
+	return summarizeMessage, nil
+}
+
+func getLastLinePosition(patch string) int {
+	lines := strings.Split(patch, "\n")
+	position := 0
+	for _, line := range lines {
+		if strings.HasPrefix(line, "@@") {
+			continue
+		}
+		position++
+	}
+	return position
+}
+
+func submitReviewComment(owner, repo string, prNumber int, commitID, filePath, comment string, position int) error {
+	client := &http.Client{}
+
+	commentRequest := CommentRequest{
+		Body:     comment,
+		Path:     filePath,
+		Position: position,
+		CommitID: commitID,
+	}
+
+	jsonData, err := json.Marshal(commentRequest)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest("POST", fmt.Sprintf(submitCommentsURL, apiBaseurl, owner, repo, prNumber), bytes.NewBuffer(jsonData))
+	if err != nil {
+		log.Println("Error creating request:", err)
+		return err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", "token "+token)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Println("Error submitting review comment:", err)
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusCreated {
+		log.Println("Review comment submitted successfully")
+	} else {
+		return fmt.Errorf("error: %d", resp.StatusCode)
+	}
+
+	return nil
+}
```

Comment:
Overall, the code patch looks relatively simple and straightforward. Below are some potential improvements and security considerations:

Improvements:
- Consider adding more error handling for each step of the process (e.g. checking if the client is nil before creating the request, checking if the response body is nil before closing it)
- Instead of hardcoding the content type and authorization header values, consider making them configurable or using constants.
- Consider adding more comments to explain what each step does and why it's necessary.

Security:
- Make sure the token value is secure and not hardcoded in the code itself (e.g. use environment variables or a secure secrets management system).
- Consider adding additional authentication and authorization checks to ensure that the requester has the appropriate permissions to make this request.
- If the request includes sensitive data, consider encrypting it before sending it to the server.